src/interface/swarm.cpp

```diff
@@ -317,28 +354,537 @@ void Swarm::Defrag() {
+              }
+              if (z > swarm_d.z_max_global_) {
+                z = swarm_d.z_min_global_ + (z - swarm_d.z_max_global_);
+              }
+            }
+          }
+        }
+      });
+
+  // Count all the particles that are Active and Not on this block, if nonzero,
+  // copy into buffers (if no send already for that buffer) and send
+
+  RemoveMarkedParticles();
+
+  vbswarm->Send(phase);
+  return true;
+}
+
+template <typename T>
+vpack_types::SwarmVarList<T> Swarm::MakeVarListAll_(ParticleVariableVector<T> variables) {
+  int size = 0;
+  vpack_types::SwarmVarList<T> vars;
+
+  for (auto it = variables.rbegin(); it != variables.rend(); ++it) {
+    auto v = *it;
+    vars.push_front(v);
+    size++;
+  }
+  return vars;
+}
+
+SwarmVariablePack<Real> Swarm::PackAllVariablesReal(PackIndexMap &vmap) {
+  std::vector<std::string> names;
+  for (auto &v : realVector_) {
+    names.push_back(v->label());
+  }
+  return PackVariablesReal(names, vmap);
+}
+
+SwarmVariablePack<int> Swarm::PackAllVariablesInt(PackIndexMap &vmap) {
+  std::vector<std::string> names;
+  for (auto &v : intVector_) {
+    names.push_back(v->label());
+  }
+  return PackVariablesInt(names, vmap);
+}
+
+SwarmVariablePack<Real> Swarm::PackVariablesReal(const std::vector<std::string> &names,
+                                                 PackIndexMap &vmap) {
+  vpack_types::SwarmVarList<Real> vars = MakeVarListAll_<Real>(realVector_);
+
+  auto pack = MakeSwarmPack<Real>(vars, &vmap);
+  SwarmPackIndxPair<Real> value;
+  value.pack = pack;
+  value.map = vmap;
+  return pack;
+}
+SwarmVariablePack<int> Swarm::PackVariablesInt(const std::vector<std::string> &names,
+                                               PackIndexMap &vmap) {
+  vpack_types::SwarmVarList<int> vars = MakeVarListAll_(intVector_);
+
+  auto pack = MakeSwarmPack<int>(vars, &vmap);
+  SwarmPackIndxPair<int> value;
+  value.pack = pack;
+  value.map = vmap;
+  return pack;
+}
+
+bool Swarm::Receive(BoundaryCommSubset phase) {
+  // Ensure all local deep copies marked BoundaryStatus::completed are actually received
+  GetBlockPointer()->exec_space.fence();
+  auto pmb = GetBlockPointer();
+
+  // Populate buffers
+  vbswarm->Receive(phase);
+
+  // Copy buffers into swarm data on this proc
+  const int maxneighbor = vbswarm->bd_var_.nbmax;
+  int total_received_particles = 0;
+  std::vector<int> neighbor_received_particles(maxneighbor);
+  for (int n = 0; n < maxneighbor; n++) {
+    if (vbswarm->bd_var_.flag[pmb->pbval->neighbor[n].bufid] == BoundaryStatus::arrived) {
+      PARTHENON_DEBUG_REQUIRE(vbswarm->recv_size[n] % vbswarm->particle_size == 0,
+                              "Receive buffer is not divisible by particle size!");
+      neighbor_received_particles[n] = vbswarm->recv_size[n] / vbswarm->particle_size;
+      total_received_particles += neighbor_received_particles[n];
+    } else {
+      neighbor_received_particles[n] = 0;
```

Comment:
Again, this method is too long, I think there is a lot of repetition here with the Send method, maybe you can reuse some of the logic.

Reply:
I separated out the buffer unloading here like in `Send`. I already removed the physical boundary conditions from `Send` so that part of the code isn't duplicated any more, I think that was the worst offender -- at least, I'm not sure where else I could share code.